### PR TITLE
✨ Add mock client and 2 tests for networking package

### PR DESCRIPTION
**Pull request by macaptain**
Refactors the Service in the networking package to use a client with CRUD functions for manipulating OpenStack resources. This interface is combined with mockgen code generation to make the package unit testable.

Two tests are added to illustrate how writing tests for the package will work. The list / filter networks and subnets functions are also slightly simplified.

**What this PR does / why we need it**: 

A refactor of the code to allow unit tests to be written for code that uses gophercloud to interface with OpenStack.

**Which issue(s) this PR fixes**:
Goes some way to fixing #843.

**Special notes for your reviewer**: 

If you like this change, my plan is to move some of the 'ports' logic to the networking package and write unit tests for it, and then apply the same changes, adding mock clients for the Service in the compute package.

I tried a few things out for making the code unit testable, including using gomega's ghttp server for mocking (too much JSON), or trying to write smaller interfaces (ended up being a lot of boiler plate and manual writing of mocks), and ended up settling on this. It's a similar pattern to what's used in cluster-api-provider-aws (see instance_test.go and the generated mocks: https://github.com/kubernetes-sigs/cluster-api-provider-aws/tree/main/pkg/cloud/services/ec2), as well as using the gomega library for matching (as recommended in the Cluster API book section on Testing.)

**TODOs**:

- [x] squashed commits
- if necessary:
  - [ ] includes documentation
  - [x] adds unit tests

/hold

### CODE REVIEW

**Review by lentzi90 — APPROVED** [review]
This looks great!

**Comment by macaptain** [comment]
This PR is two commits to make it a bit easier to review. After rebasing for the metrics changes in #932, I decided to move other metrics into the Client implementation as well for consistency in a second commit. I can squash the commits before merging if desired.

**Comment by jichenjc** [comment]
/lgtm

I think it's great PR , @hidekazuna can you take a look so that we can merge it soon (to avoid various conflict) 
and if any followup needed we can anyway do it later on

File: pkg/cloud/services/networking/client.go [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+		return nil, err
+	}
+	return ports.ExtractPorts(allPages)
+}
+
+func (c networkClient) CreatePort(opts ports.CreateOptsBuilder) (*ports.Port, error) {
+	return ports.Create(c.serviceClient, opts).Extract()
+}
+
+func (c networkClient) DeletePort(id string) error {
```
Comment (jichenjc) [inline pkg/cloud/services/networking/client.go:172]: I'd suggest we consider add metrics for those kind of API call as well , but no need to be in this PR as this is a pure refactory

File: pkg/cloud/services/networking/client.go [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package networking
+
+//go:generate mockgen -destination=mock_client/client.go -package=mock_client sigs.k8s.io/cluster-api-provider-openstack/pkg/cloud/services/networking NetworkClient
+//go:generate /usr/bin/env bash -c "cat ../../../../hack/boilerplate/boilerplate.generatego.txt mock_client/client.go > mock_client/_client.go && mv mock_client/_client.go mock_client/client.go"
```
Comment (hidekazuna) [inline pkg/cloud/services/networking/client.go:20]: CAPA and CAPZ write these two line in mock_xx/doc.go. 

- https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/pkg/cloud/services/ec2/mock_ec2iface/doc.go
- https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/azure/services/subnets/mock_subnets/doc.go

Let's follow them which means creating new file pkg/cloud/services/networking/mock_networking/doc.go and write two lines into it.

File: pkg/cloud/services/networking/mock_client/client.go [hunk 3]
```diff
@@ -0,0 +1,605 @@
+/*
```
Comment (hidekazuna) [inline pkg/cloud/services/networking/mock_client/client.go:1]: If we follow CAPA and CAPZ, this file is in pkg/cloud/services/networking/mock_networking.

File: pkg/cloud/services/networking/client.go (hunk 2 above)
Comment (macaptain) [reply pkg/cloud/services/networking/client.go]: Thanks! I agree with keeping this the same as CAPA and CAPZ. I've updated the package name and moved the `//go:generate` to doc.go. I've also squashed commits.

**Comment by hidekazuna** [comment]
Can we update like following?

networking/
├── client.go
├── floatingip.go
├── floatingip_test.go
├── mock_networking
│   ├── client.go         -> rename to client_mock.go
│   ├── doc.go
│   └── mock_client   -> delete.
│       └── client.go    -> delete. 
.......

Because CAPZ is the following.

 subnets/
├── client.go
├── mock_subnets
│   ├── client_mock.go
│   ├── doc.go
│   └── subnets_mock.go
├── subnets.go
└── subnets_test.go

**Comment by macaptain** [comment]
> Can we update like following?

Yes, of course. I've updated the folder structure and filenames to match. The mock package is called mock_networking and the generated file is client_mock.go, and mock_client and mock_client.go have been removed.

**Comment by hidekazuna** [comment]
@macaptain Thanks updating.

/approve

**Pull request merged**